core/src/sumeragi/main_loop.rs

```diff
@@ -669,445 +669,285 @@ pub fn run<F>(
-                    Ok(committed_block) => {
-                        info!(
-                            voting_block_hash=%block.hash(),
-                            "Block reached required number of votes",
-                        );
+        if let Some(VotingBlock { voted_at, .. }) = voting_block {
+            let commit_deadline = voted_at + sumeragi.commit_time;
 
-                        sumeragi.broadcast_packet(
-                            MessagePacket::new(
-                                view_change_proof_chain.clone(),
-                                BlockCommitted::from(block).into(),
-                            ),
-                            &state.current_topology,
-                        );
-                        commit_block(sumeragi, committed_block.into(), &mut state);
-                    }
-                    Err(err) => {
-                        trace!(?err, "Not enough signatures. Will be waiting for more.");
-                        voting_block_option = Some(VotingBlock { block, voted_at });
-                    }
-                }
+            if Instant::now() > commit_deadline && !view_change_suggested {
+                suggest_view_change(
+                    sumeragi,
+                    &state,
+                    &mut view_change_proof_chain,
+                    current_view_change_index,
+                );
 
-                if Instant::now() > instant_at_which_we_should_have_committed {
-                    trace!("Suspecting validating peers for not voting for block.");
-                    request_view_change(
-                        sumeragi,
-                        &mut state,
-                        &mut view_change_proof_chain,
-                        current_view_change_index,
-                    );
-                    instant_at_which_we_should_have_committed += sumeragi.commit_time;
-                }
+                view_change_suggested = true;
+            }
+        }
+    }
+}
+
+fn add_signatures<const EXPECT_VALID: bool>(
+    block: &mut VotingBlock,
+    signatures: impl IntoIterator<Item = SignatureOf<SignedBlock>>,
+) {
+    for signature in signatures {
+        if let Err(err) = block.block.add_signature(signature) {
+            let err_msg = "Signature not valid";
+
+            if EXPECT_VALID == true {
+                error!(?err, err_msg);
+            } else {
+                debug!(?err, err_msg);
             }
         }
     }
 }
 
+fn vote_for_block<F: FaultInjection>(
+    sumeragi: &SumeragiWithFault<F>,
+    state: &State,
+    mut block: VersionedCandidateBlock,
+) -> Option<VotingBlock> {
+    let addr = &sumeragi.peer_id.address;
+    let role = state.current_topology.role(&sumeragi.peer_id);
+    trace!(%addr, %role, block_hash=%block.hash(), "Block received, voting...");
+
+    if state
+        .current_topology
+        .filter_signatures_by_roles(&[Role::Leader], block.retain_verified_signatures())
```

Comment:
do you think there can be any problem if unverified signatures are removed from the block?

Reply:
This should be fine because we don't really do multisignature yet. When we do the mechanism is going to be different, so we can safely ignore it.